### PR TITLE
[TT-11244] Custom domain regex causing problems with servers

### PR DESCRIPTION
**Pull request by shults**
### **User description**
<details open>
  <summary><a href="https://tyktech.atlassian.net/browse/TT-11244" title="TT-11244" target="_blank">TT-11244</a></summary>
  <br />
  <table>
    <tr>
      <th>Summary</th>
      <td>Custom domain regex causing problems with servers</td>
    </tr>
    <tr>
      <th>Type</th>
      <td>
        Bug
        Bug
      </td>
    </tr>
    <tr>
      <th>Status</th>
      <td>In Dev</td>
    </tr>
    <tr>
      <th>Points</th>
      <td>N/A</td>
    </tr>
    <tr>
      <th>Labels</th>
      <td><a href="https://tyktech.atlassian.net/issues?jql=project%20%3D%20TT%20AND%20labels%20%3D%20codilime_refined%20ORDER%20BY%20created%20DESC" title="codilime_refined">codilime_refined</a></td>
    </tr>
  </table>
</details>

---

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning why it's required
- [ ] I would like a code coverage CI quality gate exception and have explained why

___

### **PR Type**
Bug fix, Enhancement, Tests

___

### **Description**
- Fixed handling of custom domain regex in OAS server URLs.

- Added robust parsing for server URLs with regex and variables.

- Introduced `RemoveServer` method to remove servers by normalized URL.

- Enhanced and expanded tests for server URL parsing and server management.

___

### Diagram Walkthrough

```mermaid
flowchart LR
  oasutil["oasutil/servers.go: Server URL parser & helpers"]
  oas["apidef/oas/oas.go: AddServers/RemoveServer logic"]
  oastest["apidef/oas/oas_test.go: AddServers & RemoveServer tests"]
  gateway["gateway/api.go: Integrate AddServers error handling"]
  oasutiltest["internal/oasutil/servers_test.go: Parser unit tests"]

  oasutil -- "used by" --> oas
  oas -- "tested by" --> oastest
  oasutil -- "tested by" --> oasutiltest
  gateway -- "calls" --> oas
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>oas.go</strong><dd><code>Refactor and fix OAS server URL add/remove logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apidef/oas/oas.go

<ul><li>Added <code>RemoveServer</code> method for removing servers by normalized URL.<br> <li> Refactored <code>AddServers</code> to use new server URL parser and handle <br>regex/variables.<br> <li> Updated <code>AddServers</code> to return errors on invalid input.<br> <li> Improved handling of server variables and normalization.</ul>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/7233/files#diff-80279b1d59499a41a77ff7a16a6e2c9b9b785a4fd1326c351da6884c867658d7">+36/-7</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>servers.go</strong><dd><code>Add robust OAS server URL parser utility</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/oasutil/servers.go

<ul><li>Introduced server URL parser handling regex and variables.<br> <li> Added error types for parsing issues and collisions.<br> <li> Provided normalization and variable extraction logic.<br> <li> Exported <code>ParseServerUrl</code> for external use.</ul>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/7233/files#diff-98dd06199bf9992e099563df9150f18cb38094f4dae3299f33c5330722ddac3d">+162/-0</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>oas_test.go</strong><dd><code>Expand and improve OAS server management tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apidef/oas/oas_test.go

<ul><li>Updated <code>TestOAS_AddServers</code> to test new server URL parsing and <br>normalization.<br> <li> Added <code>Test_RemoveServer</code> for new removal logic and edge cases.<br> <li> Improved assertions for server variables and error handling.</ul>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/7233/files#diff-74029ee88132d30d6478c96a35f8bb2200e0c8e6f42f2c9b147dc6bb7ce74644">+103/-42</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>servers_test.go</strong><dd><code>Add unit tests for OAS server URL parser</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/oasutil/servers_test.go

<ul><li>Added comprehensive unit tests for server URL parser.<br> <li> Covered positive and negative cases, including regex and collisions.</ul>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/7233/files#diff-4a274e0f05feb520d7ff68fb48a7a28020ea7525820ac8105f4c898854a19af8">+91/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Bug fix</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>api.go</strong><dd><code>Integrate AddServers error handling in API creation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

gateway/api.go

<ul><li>Updated OAS API add handler to handle errors from <code>AddServers</code>.<br> <li> Returns HTTP 400 if server URL parsing fails.</ul>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/7233/files#diff-644cda3aeb4ac7f325359e85fcddb810f100dd5e6fa480b0d9f9363a743c4e05">+4/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: internal/oasutil/servers.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+type serverVariable struct {
+	name, pattern string
+}
+
+func (p *serverUrlParser) parse(url string) (*ServerUrl, error) {
+	p.url = url
+
+	result := new(ServerUrl)
```
Comment (MaciekMis) [inline internal/oasutil/servers.go]: create factory function for ServerUrl that initializes default values for variables

**Review by MaciekMis — CHANGES_REQUESTED** [review]

File: apidef/oas/oas.go [hunk 2]
```diff
@@ -340,28 +343,51 @@ func (s *OAS) getTykOperations() (operations Operations) {
 	return
 }
 
+func (s *OAS) RemoveServer(serverUrl string) error {
```
Comment (edsonmichaque) [inline apidef/oas/oas.go:348]: What do you think about having godocs here?
